mono/utils/mono-threads.c

```diff
@@ -1835,3 +1911,181 @@ mono_thread_info_get_tools_data (void)
+
+	// For synch calls
+	gboolean *finished;
+	MonoSemType *response_sem;
+
+	// Variably-sized, size is thread->payload_size
+	gpointer payload [MONO_ZERO_LEN_ARRAY]; 
+} UtilityThreadQueueEntry;
+
+static void
+free_queue_entry (gpointer p)
+{
+	UtilityThreadQueueEntry *util = (UtilityThreadQueueEntry *) p;
+	mono_lock_free_free (p, util->thread->message_block_size);
+}
+
+static gboolean
+utility_thread_handle_inbox (MonoUtilityThread *thread)
+{
+	UtilityThreadQueueEntry *entry = (UtilityThreadQueueEntry *) mono_lock_free_queue_dequeue (&thread->work_queue);
+	if (!entry)
+		return FALSE;
+	
+	thread->callbacks.command (thread->state_ptr, &entry->payload);
+	if (entry->response_sem) {
+		*entry->finished = TRUE;
+		mono_os_sem_post (entry->response_sem);
+	}
+
+	mono_thread_hazardous_try_free (entry, free_queue_entry);
+
+	return TRUE;
+}
+
+static void *
+utility_thread (void *arg)
+{
+	mono_thread_info_wait_inited ();
+	mono_thread_info_attach ();
+
+	MonoUtilityThread *thread = (MonoUtilityThread *) arg;
+	thread->callbacks.init (&thread->state_ptr);
+
+	while (mono_atomic_load_i32 (&thread->run_thread)) {
+		MONO_ENTER_GC_SAFE;
+		mono_os_sem_timedwait (&thread->work_queue_sem, 1000, MONO_SEM_FLAGS_NONE);
+		MONO_EXIT_GC_SAFE;
+		utility_thread_handle_inbox (thread);
+	}
+
+	/* Drain any remaining entries on shutdown. */
+	while (utility_thread_handle_inbox (thread));
+
+	mono_os_sem_destroy (&thread->work_queue_sem);
+
+	thread->callbacks.cleanup (thread->state_ptr);
+
+	return NULL;
+}
+
+MonoUtilityThread *
+mono_utility_thread_launch (size_t payload_size, MonoUtilityThreadCallbacks *callbacks)
+{
+	MonoUtilityThread *thread = g_malloc0 (sizeof (MonoUtilityThread));
+	size_t entry_size = offsetof (UtilityThreadQueueEntry, payload) + payload_size;
+
+	thread->message_block_size = mono_pagesize ();
+	thread->payload_size = payload_size;
+	thread->callbacks = *callbacks;
+
+	mono_lock_free_queue_init (&thread->work_queue);
+	mono_lock_free_allocator_init_size_class (&thread->message_size_class, entry_size, thread->message_block_size);
+	mono_lock_free_allocator_init_allocator (&thread->message_allocator, &thread->message_size_class, MONO_MEM_ACCOUNT_OTHER);
+	mono_os_sem_init (&thread->work_queue_sem, 0);
+	mono_atomic_store_i32 (&thread->run_thread, 1);
+
+	if (!mono_native_thread_create (&thread->thread_id, utility_thread, thread))
+		g_error ("Could not create utility thread");
+
+	return thread;
+}
+
+static void
+mono_utility_thread_send_internal (MonoUtilityThread *thread, UtilityThreadQueueEntry *entry)
+{
+	mono_lock_free_queue_node_init (&entry->node, FALSE);
+	mono_lock_free_queue_enqueue (&thread->work_queue, &entry->node);
+	mono_os_sem_post (&thread->work_queue_sem);
+}
+
+void
+mono_utility_thread_send (MonoUtilityThread *thread, gpointer message)
+{
+	int small_id = mono_thread_info_get_small_id ();
+	if (small_id < 0) {
+#if 1
+		fprintf (stderr, "Dropping message send because thread not attached yet\n");
```

Comment:
Shouldn't these be `g_error`s?

Reply:
g_error will try to go through the logger, so I've tried to avoid using them in this machinery. I can figure something out if people would prefer to see that here.

Reply:
`g_printerr`?  so it works for embedders too.